File: p1.rb
a=eval(gets.chomp.tr('(','[').tr(')',']'))
b=2+a[0][1]
for c in a.map{|z|3+z[1]+z[2]}.max.downto(1)
  for d in a
    e=3+d[1]+d[2]
    if c>e
      print ' '*(3+d[0]+d[1])
    elsif c>b&&c>2+d[2]
      print ' '*(d[1]+1-(e-c))
    elsif c<=b&&c>2+ d[2]
      print '_'*(c-2-d[2])
    end
    if c==3+d[1]+d[2]
      print '+'+'-'*d[0]+'+'
    elsif c<3+d[1]+d[2]&&c>2+d[2]
      print '/'+' '*d[0]+'/'
    elsif c==2+d[2]
      print '+'+'-'*d[0]+'+'
    elsif c<2+d[2]&&c>1
      print '|'+' '*d[0]+'|'
    elsif c==1
      print '+'+'-'*d[0]+'+'
    end
    if c<e&&c>d[1]+2
      print ' '*[(e-1-c),d[2],d[1]].min+'|'
    elsif c==d[1]+2
      print ' '*[d[1],d[2],d[1]].min+'+'
    elsif c<d[1]+2&&c>1
      print ' '*[(c-2),d[2],d[1]].min+'/'
    end
    if c<=b
      print '_'*(b-c+1)
    else
      print ' '
    end
  end
  puts
end

File: p1.py
l=input()
c=2+l[0][1]
x=' '
for n in range(max([3+b[1]+b[2]for b in l]),0,-1):
	o=''
	for z in l:
		v=z[0]
		f=z[1]
		p=z[2]
		q=min([f,p])
		h=3+f+p
		s=2+p
		t=2+f
		if n>h:
			o+=x*(3+v+f)
		elif n>s:
			o+=x*(n-s) if n>c else '_'*(n-s)
		if n==h or n==s or n==1:
			o+='+'+'-'*v+'+'
		elif n<h and n>s:
			o+='/'+x*v+'/'
		elif n<s and n>1:
			o+='|'+x*v+'|'
		if n<h and n>t:
			o+=x*min((h-1-n),q)+'|'
		elif n==t:
			o+=x*q+'+'
		elif n<t and n>1:
			o+=x*min((n-2),q)+'/'
		o+='_'*(c-n+1)if n<=c else x
	print o
